include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -891,6 +895,267 @@ namespace VectorTools
+          break;
+
+        case 6:
+          dealii::VectorTools::project_generic<components, 6>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 7:
+          dealii::VectorTools::project_generic<components, 7>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 8:
+          dealii::VectorTools::project_generic<components, 8>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        default:
+          Assert(false, ExcNotImplemented());
+        }
+    }
+
+
+
+    // Helper interface for the matrix-free implementation of project().
+    // Used to determine the number of components.
+    template <int dim, typename VectorType, int spacedim>
+    void project_parallel (const Mapping<dim, spacedim>           &mapping,
+                           const DoFHandler<dim,spacedim>         &dof,
+                           const ConstraintMatrix                 &constraints,
+                           const Quadrature<dim>                  &quadrature,
+                           const Function<spacedim, typename VectorType::value_type> &function,
+                           VectorType                             &vec_result,
+                           const bool                             enforce_zero_boundary,
+                           const Quadrature<dim-1>                &q_boundary,
+                           const bool                             project_to_boundary_first)
+    {
+      switch (dof.get_fe().n_components())
+        {
+        case 1:
+          project_parallel_helper<1>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 2:
+          project_parallel_helper<2>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 3:
+          project_parallel_helper<3>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 4:
+          project_parallel_helper<4>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        default:
+          Assert(false, ExcNotImplemented());
+        }
+    }
+
+
+
+    template <int dim, typename VectorType>
+    void project_matrix_free (const Mapping<dim>            &mapping,
```

Comment:
this name is mis-leading here as you choose between matrix-free and old implementation. Since this is called only for dim==spacedim, maybe call it `project_dim_eq_spacedim`